### PR TITLE
[VBLOCKS-1352] Incoming call registration

### PR DESCRIPTION
**Pull request by mhuynh5757**
This PR introduces new application bootstrapping actions. Within the bootstrapping action, Twilio Voice access token and registration occurs.

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: app/src/store/bootstrap.ts
Comment (kpchoy) [inline app/src/store/bootstrap.ts:20]: Why do we need to "bootstrapApp" now? What was wrong with the previous login implementation?

File: app/src/store/voice/accessToken.ts
Comment (kpchoy) [inline app/src/store/voice/accessToken.ts:77]: I don't think this "slice" comment is necessary since line 79 says `AsyncStoreSlice`

File: app/src/store/voice/call/outgoingCall.ts
Comment (kpchoy) [inline app/src/store/voice/call/outgoingCall.ts:50]: Should we add `if (error)` to these cases as well?

File: server/src/routes/twiml.ts
Comment (kpchoy) [inline server/src/routes/twiml.ts:20]: Why log instead of error these messages? Aren't these all failed requests?

**Review by kpchoy — CHANGES_REQUESTED** [review]
Can you make sure the Android Detox CI passes before merging

**Comment by charliesantos** [comment]
Great stuff @mhuynh5757 . I like the improvements. These will definitely help us and our customers in the future. Arguably, some of the changes are probably not required for the completion of the story. As a note for the future, let's draw the line between what is required now and what can be pushed out for later to prevent unnecessary delays and additional work for the team. Me and @kpchoy also needs to spend some time to review 😉. But overall, good work! I will do a deeper review and provide feedback. Just need more time to finish. Also, let's merge @kpchoy 's PR first and make sure your PR is green.

File: app/src/screens/SignIn/index.tsx
Comment (charliesantos) [reply app/src/screens/SignIn/index.tsx:70]: +1 to @kpchoy 's comment.

Comment (mhuynh5757) [reply app/src/screens/SignIn/index.tsx:70]: Yep, still works.

File: app/src/screens/StackNavigator.tsx
Comment (mhuynh5757) [reply app/src/screens/StackNavigator.tsx:21]: Good point, but I think it should be part of the UI/UX tickets. This is temporary for now.

File: app/src/store/bootstrap.ts
Comment (mhuynh5757) [reply app/src/store/bootstrap.ts:20]: Having a bootstrapping action allows us to place more complex logic into our app and have it start somewhere outside of a component. I issue this action in the `App.tsx` file before even rendering the app, and so by the time the rendering is done we can already skip the login screen if the user was already logged in.

File: app/src/store/voice/call/outgoingCall.ts
Comment (mhuynh5757) [reply app/src/store/voice/call/outgoingCall.ts:50]: These events should only ever be emitted when there are errors. Meaning that those errors are never `undefined` and are always truthy. The only exception is the `Disconnected` event where the listener param is `error?` since a disconnect can occur with or without an error.

File: server/src/routes/twiml.ts
Comment (mhuynh5757) [reply server/src/routes/twiml.ts:20]: This is a good question, I would say errors should be emitted if they are the fault of the server code, but in this case it's the fault of the request. What do you think? cc @charliesantos

File: app/src/screens/StackNavigator.tsx
Comment (charliesantos) [reply app/src/screens/StackNavigator.tsx:21]: Which ticket is this part of? Let's make sure it's captured or else we'll forget this.

File: app/src/store/__tests__/registration.test.ts
Comment (charliesantos) [inline app/src/store/__tests__/registration.test.ts:24]: Ha, so this is the middleware we talked about? I don't have preference between this and jest mocks. I'll leave it up to you.

File: app/src/store/bootstrap.ts
Comment (charliesantos) [reply app/src/store/bootstrap.ts:20]: The previous login implementation is also good, but I think this is a good improvement.

File: server/src/routes/twiml.ts
Comment (charliesantos) [reply server/src/routes/twiml.ts:20]: We need to add loglevels and define what each level means e.g. info, warn, debug, error. For now, let us use `log` for consistency.

File: server/src/utils/auth.ts
Comment (charliesantos) [inline server/src/utils/auth.ts:21]: Don't make this specific to axios. We're basically exposing what library we're using for API calls. This should be an error that we define. Maybe `AUTH_ERROR`?

File: server/src/utils/auth.ts
Comment (charliesantos) [inline server/src/utils/auth.ts:38]: What is inside `error` object? This is coming from axios right? Can we extract only the information that is needed?

**Review by charliesantos — CHANGES_REQUESTED** [review]

File: app/src/screens/StackNavigator.tsx
Comment (mhuynh5757) [reply app/src/screens/StackNavigator.tsx:21]: Added a note to this ticket https://issues.corp.twilio.com/browse/VBLOCKS-1349

File: app/.gitignore
Comment (mhuynh5757) [reply app/.gitignore:73]: Yes, it's for building the app. The Android SDK uses Firebase Cloud Messaging for push notifications and this file is required. It's generated when making a Firebase Android App as part of setting up a Firebase project.

File: app/src/App.tsx
Comment (mhuynh5757) [reply app/src/App.tsx:11]: Redux toolkit async action creators automatically put an `AbortController` in the returned `Promise` when dispatching them. Reading up on the Redux toolkit docs recommends this so the `bootstrapApp` action stops dispatching further actions if `AbortController` is used to abort the thunk. The return value of `useEffect` is invoked when the component is dismounted, which in this case should really never happen since it's the root component but just in case.

File: app/src/screens/Dialer/hooks.ts
Comment (mhuynh5757) [reply app/src/screens/Dialer/hooks.ts:122]: These are duplicate lines, there's some above logic that returns on dispatch-rejection.

File: app/src/store/__tests__/registration.test.ts
Comment (mhuynh5757) [reply app/src/store/__tests__/registration.test.ts:24]: Yep, this is it. Super simple so I'm not _too_ concerned about complicating actual app logic, and it helps me test tremendously.

File: app/src/store/__tests__/store.test.ts
Comment (mhuynh5757) [reply app/src/store/__tests__/store.test.ts:37]: Haha, yeah, I keep forgetting about this one. Will change.

File: app/src/store/app.ts
Comment (mhuynh5757) [reply app/src/store/app.ts:7]: We will still be able to differentiate since the actual `action.type` value is also logged. For example, the logs will look something like this:
```
voice/getAccessToken/pending
voice/getAccessToken/rejected SOME_ERROR_PAYLOAD
```

**Review by kpchoy — APPROVED** [review]

File: app/src/store/user.ts
Comment (mhuynh5757) [reply app/src/store/user.ts:110]: Good point, should be never. But as I was looking at all of our async action creators, I wanted them all to follow a similar pattern as such:

```ts
type RejectValue =
  | {
      reason: 'foo';
      error: SerializedError;
    }
  | {
      reason: 'bar';
    };
```

And I wanted to change all of our async creators to this pattern, but I think it should go into another ticket/PR since this one is so bloated already.

So making the error optional here for the `ID_TOKEN_UNDEFINED` reason is the smallest change I could think of that still is technically correct just for the sake of this PR. Ideally, the ticket I just mentioned is to clean up all the action creators so they follow the same pattern.

File: server/src/utils/auth.ts
Comment (mhuynh5757) [reply server/src/utils/auth.ts:38]: To be honest, I looked at the Auth0 API docs and they're not very clear about the things we can expect from their API, error or success. The payloads they return are not clearly documented.

Comment (mhuynh5757) [reply server/src/utils/auth.ts:38]: I'm probably not looking in the right place, though.

File: app/.gitignore
Comment (mhuynh5757) [reply app/.gitignore:73]: Here's a ticket for it https://issues.corp.twilio.com/browse/VBLOCKS-1631

File: app/src/App.tsx
Comment (charliesantos) [reply app/src/App.tsx:11]: Can you put a comment above this line that explains it? It's not immediately clear.

File: server/src/utils/auth.ts
Comment (charliesantos) [reply server/src/utils/auth.ts]: Can you try triggering an error and log it to see how it looks? We'll be able to see the error format that way and only grab the info we need, not the whole thing.

File: app/src/store/__tests__/store.test.ts
Comment (charliesantos) [inline app/src/store/__tests__/store.test.ts:37]: >it('should make...

File: server/src/utils/auth.ts
Comment (charliesantos) [reply server/src/utils/auth.ts]: I see you addressed this. Resolving.

**Review by charliesantos — CHANGES_REQUESTED** [review]
I left a few more responses @mhuynh5757 .

**Review by charliesantos — APPROVED** [review]

**Pull request merged**
